packages/ckeditor5-list/src/documentlist/converters.js

```diff
@@ -0,0 +1,521 @@
+					findAndAddListHeadToMap( entry.range.start, itemToListHead );
+
+					if ( entry.attributeNewValue === null ) {
+						findAndAddListHeadToMap( entry.range.start.getShiftedBy( 1 ), itemToListHead );
+						refreshItemParagraphIfNeeded( item, [] );
+					} else {
+						changedItems.add( item );
+					}
+				} else if ( item.hasAttribute( 'listItemId' ) ) {
+					refreshItemParagraphIfNeeded( item );
+				}
+			}
+		}
+
+		for ( const listHead of itemToListHead.values() ) {
+			checkList( listHead );
+		}
+
+		for ( const item of itemsToRefresh ) {
+			editing.reconvertItem( item );
+		}
+
+		function checkList( listHead ) {
+			const visited = new Set();
+			const stack = [];
+
+			for (
+				let prev = null, item = listHead;
+				item && item.hasAttribute( 'listItemId' );
+				prev = item, item = item.nextSibling
+			) {
+				if ( visited.has( item ) ) {
+					continue;
+				}
+
+				const itemIndent = item.getAttribute( 'listIndent' );
+
+				if ( prev && itemIndent < prev.getAttribute( 'listIndent' ) ) {
+					stack.length = itemIndent + 1;
+				}
+
+				stack[ itemIndent ] = {
+					id: item.getAttribute( 'listItemId' ),
+					type: item.getAttribute( 'listType' )
+				};
+
+				const blocks = getListItemElements( item, 'forward' );
+
+				for ( const block of blocks ) {
+					visited.add( block );
+
+					refreshItemParagraphIfNeeded( block, blocks );
+					refreshItemWrappingIfNeeded( block, stack );
+				}
+			}
+		}
+
+		function refreshItemParagraphIfNeeded( item, blocks ) {
+			if ( !item.is( 'element', 'paragraph' ) ) {
+				return;
+			}
+
+			const viewElement = editing.mapper.toViewElement( item );
+
+			if ( !viewElement ) {
+				return;
+			}
+
+			const useBogus = shouldUseBogusParagraph( item, blocks );
+
+			if ( useBogus && viewElement.is( 'element', 'p' ) ) {
+				itemsToRefresh.add( item );
+			} else if ( !useBogus && viewElement.is( 'element', 'span' ) ) {
+				itemsToRefresh.add( item );
+			}
+		}
+
+		function refreshItemWrappingIfNeeded( item, stack ) {
+			// Items directly affected by some "change" don't need a refresh, they will be converted by their own changes.
+			if ( changedItems.has( item ) ) {
+				return;
+			}
+
+			const viewElement = editing.mapper.toViewElement( item );
+			let stackIdx = stack.length - 1;
+
+			for (
+				let element = viewElement.parent;
+				!element.is( 'editableElement' );
+				element = element.parent
+			) {
+				if ( isListItemView( element ) ) {
+					if ( element.id != stack[ stackIdx ].id ) {
+						break;
+					}
+				} else if ( isListView( element ) ) {
+					const expectedElementName = getViewElementNameForListType( stack[ stackIdx ].type );
+
+					if ( element.name != expectedElementName ) {
+						break;
+					}
+
+					stackIdx--;
+
+					// Don't need to iterate further if we already know that the item is wrapped appropriately.
+					if ( stackIdx < 0 ) {
+						return;
+					}
+				}
+			}
+
+			itemsToRefresh.add( item );
+		}
```

Comment:
Do we need to create all these helpers every time `change:data` kicks in? I'd extract them somewhere else.